qutebrowser/keyinput/macros.py

```diff
@@ -0,0 +1,99 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2016 Jan Verbeek (blyxxyz) <ring@openmailbox.org>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Keyboard macro system."""
+
+from qutebrowser.commands import cmdexc, cmdutils, runners
+from qutebrowser.utils import message, objreg, usertypes
+
+
+class MacroRecorder:
+    """An object for recording and running keyboard macros."""
+
+    def __init__(self):
+        self.macro = {}
+        self.recording_macro = None
+        self.macro_count = {}
+
+    @cmdutils.register(instance='macro-recorder', name='record-macro')
+    @cmdutils.argument('win_id', win_id=True)
+    @cmdutils.argument('register')
+    def record_macro_command(self, win_id, register=None):
+        """Start or stop recording a macro.
+
+        Args:
+            register: Which register to store the macro in.
+        """
+        if self.recording_macro is None:
+            if register is None:
+                mode_manager = objreg.get('mode-manager', scope='window',
+                                          window=win_id)
```

Comment:
You can use `modeman.instance(win_id)` from the `keyinput.modeman` module instead.